File: 5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start
class Solution:
    
    indices = []
    maxLength = -float("inf")
    
    def isPalindrome(self,low,high,s):
            
        while low>=0 and high<len(s) and s[low]==s[high]:
            low-=1
            high+=1

        if self.maxLength<high-low+1:
            self.maxLength = high-low+1
            self.indices = [low,high]
    
    def longestPalindrome(self, s: str) -> str:
        self.indices = []
        self.maxLength = -float("inf")
        
        if len(s) == 1:
            return s

        for i in range(len(s)):
            self.isPalindrome(i,i,s)
            
            self.isPalindrome(i,i+1,s)

        if self.indices:
            return s[self.indices[0]+1 : self.indices[1]]
        else:
            return ""

# @lc code=end

File: 5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
#include <iostream>
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
    string longestPalindrome(string s) {
        string t = "#";
        for (char c : s) {
            t += c;
            t += "#";
        }
        int n = t.size();
        vector<int> p(n);
        int c = 0, r = 0, max_len = 0, center = 0;
        for (int i = 0; i < n; ++i) {
            if (i < r) {
                p[i] = min(r - i, p[2 * c - i]);
            }
            while (i + p[i] + 1 < n && i - p[i] - 1 >= 0 && t[i + p[i] + 1] == t[i - p[i] - 1]) {
                ++p[i];
            }
            if (i + p[i] > r) {
                c = i;
                r = i + p[i];
            }
            if (p[i] > max_len) {
                max_len = p[i];
                center = i;
            }
        }
        return s.substr((center - max_len) / 2, max_len);
    }
};

// @lc code=end

